File: solutions/python/bit-manipulation/maximum-xor-of-two-numbers-in-an-array.py
# LeetCode 421 — Maximum XOR of Two Numbers in an Array (Medium)
# Category: Bit Manipulation · Approach: Greedy prefix set
# Time: O(32n) | Space: O(n)
# Source: https://leetcode.com/problems/maximum-xor-of-two-numbers-in-an-array/

def find_maximum_xor(nums: list[int]) -> int:
    max_xor = 0
    mask = 0
    for bit in range(30, -1, -1):
        mask |= (1 << bit)
        prefixes = {n & mask for n in nums}
        candidate = max_xor | (1 << bit)
        if any((candidate ^ p) in prefixes for p in prefixes):
            max_xor = candidate
    return max_xor

File: solutions/typescript/bit-manipulation/maximum-xor-of-two-numbers-in-an-array.ts
// LeetCode 421 — Maximum XOR of Two Numbers in an Array (Medium)
// Category: Bit Manipulation · Approach: Greedy prefix set
// Time: O(32n) | Space: O(n)
// Source: https://leetcode.com/problems/maximum-xor-of-two-numbers-in-an-array/

function findMaximumXOR(nums: number[]): number {
  let maxXor = 0;
  let mask = 0;
  for (let bit = 30; bit >= 0; bit--) {
    mask |= (1 << bit);
    const prefixes = new Set<number>();
    for (const n of nums) prefixes.add(n & mask);
    const candidate = maxXor | (1 << bit);
    for (const p of prefixes) {
      if (prefixes.has(candidate ^ p)) {
        maxXor = candidate;
        break;
      }
    }
  }
  return maxXor;
}
